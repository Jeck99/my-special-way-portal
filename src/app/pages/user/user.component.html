<div class="main-content">
  <div class="container-fluid">
    <div class="header-menu-container msw-table-tools-header">
      <div class="header-table-count">
        <div>
          <span class="header-table-count-label">משתמשים</span>
          <span class="header-table-count-number">({{dataSource.data.length}})</span>
        </div>
      </div>
      <div class="example-header msw-table-tools-header">
        <button mat-raised-button color="primary" (click)="addNewUser()">
          <i class="material-icons">
            add
          </i>
          הוספת משתמש חדש
        </button>
      </div>
    </div>
    <div class="row">
      <div class="col-lg-12 col-md-12 col-sm-12">
        <mat-table #table [dataSource]="dataSource" class="msw-table"
                   matSort matSortActive="firstName" matSortDisableClear matSortDirection="asc">

          <!-- ID Column -->
          <ng-container matColumnDef="name">
            <mat-header-cell *matHeaderCellDef mat-sort-header>שם</mat-header-cell>
            <mat-cell *matCellDef="let row" (click)="updateUser(row._id, row.firstname, row.lastname, row.email, row.username)"> {{row.firstname}} {{ row.lastname }}</mat-cell>
          </ng-container>

          <!-- Progress Column -->
          <ng-container matColumnDef="class">
            <mat-header-cell *matHeaderCellDef mat-sort-header>כיתה</mat-header-cell>
            <mat-cell *matCellDef="let row" (click)="updateUser(row._id, row.firstname, row.lastname, row.email, row.username)">{{(row.Class && row.Class.name) || '' }}</mat-cell>
          </ng-container>

          <!-- Progress Column -->
          <ng-container matColumnDef="username">
            <mat-header-cell *matHeaderCellDef mat-sort-header>שם משתמש</mat-header-cell>
            <mat-cell *matCellDef="let row" (click)="updateUser(row._id, row.firstname, row.lastname, row.email, row.username)">{{row.username}}</mat-cell>
          </ng-container>


          <!-- Name Column -->
          <ng-container matColumnDef="type">
            <mat-header-cell *matHeaderCellDef mat-sort-header>סוג משתמש</mat-header-cell>
            <mat-cell *matCellDef="let row" (click)="updateUser(row._id, row.firstname, row.lastname, row.email, row.username)">{{toHebrew(row.role)}}</mat-cell>
          </ng-container>

          <!-- Progress Column -->
          <ng-container matColumnDef="email">
            <mat-header-cell *matHeaderCellDef mat-sort-header>אימייל</mat-header-cell>
            <mat-cell *matCellDef="let row" (click)="updateUser(row._id, row.firstname, row.lastname, row.email, row.username)">{{row.email}}</mat-cell>
          </ng-container>

          <ng-container matColumnDef="recoverpassword">
            <mat-header-cell *matHeaderCellDef></mat-header-cell>
            <mat-cell *matCellDef="let row">
              <button mat-icon-button color="primary" style="text-decoration: underline">שחזר סיסמא</button>
            </mat-cell>
          </ng-container>

          <!--&lt;!&ndash; State Column &ndash;&gt;-->
          <ng-container matColumnDef="deleteUser">-->
            <mat-header-cell *matHeaderCellDef></mat-header-cell>
            <mat-cell *matCellDef="let row">
              <button mat-icon-button color="primary" (click)="deleteUser(row._id, row.firstname, row.lastname)"><i class="material-icons"> delete </i></button>
            </mat-cell>
          </ng-container>

          <mat-header-row *matHeaderRowDef="displayedColumns"></mat-header-row>
          <mat-row *matRowDef="let row; columns: displayedColumns;">
        </mat-row>
        </mat-table>

        <mat-paginator [pageSizeOptions]="[5, 10, 25, 100]"></mat-paginator>
      </div>
    </div>
  </div>
</div>
