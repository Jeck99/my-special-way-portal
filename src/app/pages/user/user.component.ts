import {Component, ElementRef, OnInit, ViewChild, ChangeDetectorRef, AfterViewInit} from '@angular/core';
import {MatDialog, MatPaginator, MatSort, MatTableDataSource} from '@angular/material';
import {UserService} from './services/user.service';
import {merge} from 'rxjs/observable/merge';
import {of as observableOf} from 'rxjs/observable/of';
import {catchError} from 'rxjs/operators/catchError';
import {map} from 'rxjs/operators/map';
import {startWith} from 'rxjs/operators/startWith';
import {switchMap} from 'rxjs/operators/switchMap';
import {element} from 'protractor';
import {forEach} from '@angular/router/src/utils/collection';
import {default as User, UserType} from '../../models/user.model';
import Student from '../../models/student.model';
import {AddStudentDialogComponent} from '../student/dialogs/add/add-student.dialog';
import {AddUserDialogComponent} from './dialogs/add/add-user.dialog';

@Component({
  selector: 'app-user',
  templateUrl: './user.component.html',
  styleUrls: ['./user.component.scss']
})
export class UserComponent implements OnInit, AfterViewInit {

  displayedColumns = ['name', 'userName', 'class', 'type', 'editDetails', 'deleteUser'];
  dataSource = new MatTableDataSource();
  resultsLength = 0;

  @ViewChild(MatPaginator) paginator: MatPaginator;
  @ViewChild(MatSort) sort: MatSort;
  @ViewChild('table') table: ElementRef;

  constructor(private userService: UserService,
              public dialog: MatDialog,
              private changeDetectorRefs: ChangeDetectorRef) { }

  ngOnInit(): void {

    this.sort.sortChange.subscribe(() => this.paginator.pageIndex = 0);
    merge(this.sort.sortChange, this.paginator.page)
      .pipe(
        startWith({}),
        switchMap(() => {
          if (this.userService !== undefined) {
            return this.userService.getAllUsers();
          }
        }),
        map(data => {
          this.resultsLength = data.length;
          for (const user of data) {
            user.userType = UserType[user.userType];
          }
          return data;
        }),
        catchError(() => {
          return observableOf([]);
        })
      ).subscribe(data => this.dataSource.data = data);

  }

  ngAfterViewInit() {
    this.dataSource.paginator = this.paginator;
    this.dataSource.sort = this.sort;
  }
  applyFilter(filterValue: string) {
    filterValue = filterValue.trim(); // Remove whitespace
    filterValue = filterValue.toLowerCase(); // Datasource defaults to lowercase matches
    this.dataSource.filter = filterValue;
  }

  addNewUser() {
    const dialogRef = this.dialog.open(AddUserDialogComponent, {
      data: {user: User }, height: '600px',
      width: '350px'
    });
    dialogRef.afterClosed().subscribe(data => {
      if (data) {
        // After dialog is closed we're doing frontend updates
        // For add we're just pushing a new row inside DataService
        this.dataSource.data.push(data);
        this.dataSource.paginator = this.paginator;

      }
    });
  }

  private refreshTable() {
    // if there's a paginator active we're using it for refresh
    if (this.dataSource.paginator.hasNextPage()) {
      this.dataSource.paginator.nextPage();
      this.dataSource.paginator.previousPage();
      // in case we're on last page this if will tick
    } else if (this.dataSource.paginator.hasPreviousPage()) {
      this.dataSource.paginator.previousPage();
      this.dataSource.paginator.nextPage();
      // in all other cases including active filter we do it like this
    } else {
      this.dataSource.filter = '';
      // this.dataSource.filter = this.filter.nativeElement.value;
    }
  }
}
