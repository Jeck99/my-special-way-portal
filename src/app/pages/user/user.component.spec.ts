jest.mock('./services/user.graphql.service');

import { TestBed } from '@angular/core/testing';
import { NO_ERRORS_SCHEMA } from '@angular/core';
import { MatHeaderRowDef, MatRowDef, MatHeaderRow, MatDialog, MatSort } from '@angular/material';
import { UserComponent } from './user.component';
import { UserService } from './services/user.graphql.service';
import { UserType } from '../../models/user.model';

describe('user component', () => {
  beforeEach(async () => {

    class UserDialogMock {
      open = jest.fn().mockImplementation(() => {
        return {
          afterClosed: () => {
            return {
              subscribe: jest.fn(),
            };
          },
        };
      });
    }

    // class UserServiceMock {
    //   getAllUsers = jest.fn().mockImplementation(() => {
    //     return '';
    //   });
    // }

    TestBed.configureTestingModule({
      imports: [

      ],
      declarations: [
        UserComponent,
        MatHeaderRow,
        MatRowDef,
        MatHeaderRowDef,

      ],
      providers: [
        UserService,
        { provide: MatDialog, useClass: UserDialogMock },
        // { provide: UserService, useClass: UserServiceMock },

      ],
      schemas: [NO_ERRORS_SCHEMA],
    });

    // shallow = new Shallow(UserComponent, AppModule);
  });

  it('should render component as described in snapshot', () => {
    const fixture = TestBed.createComponent(UserComponent);
    expect(fixture).toMatchSnapshot();
  });

  it('should open dialog when calling addNewUser function', () => {
    const fixture = TestBed.createComponent(UserComponent);
    fixture.componentInstance.addNewUser();
    const userDialogMock = TestBed.get(MatDialog);
    expect(userDialogMock.open).toHaveBeenCalled();
  });

  it('should open dialog when calling deleteUser function', () => {
    const fixture = TestBed.createComponent(UserComponent);
    fixture.componentInstance.deleteUser(123, 'sad', 'asd', UserType.MANAGER);
    const userDialogMock = TestBed.get(MatDialog);
    expect(userDialogMock.open).toHaveBeenCalled();
  });

  it('should open dialog when calling updateeUser function', () => {
    const fixture = TestBed.createComponent(UserComponent);
    fixture.componentInstance.updateUser(123, 'sad', 'asd', 'asd', 'asd');
    const userDialogMock = TestBed.get(MatDialog);
    expect(userDialogMock.open).toHaveBeenCalled();
  });
  // it('should call getAllUsers on init', () => {
  //   const fixture = TestBed.createComponent(UserComponent);
  //   fixture.componentInstance.sort = new MatSort();
  //   fixture.componentInstance.table = new ElementRef(MatTable);
  //   fixture.componentInstance.ngOnInit();
  //   const UserServiceMock = TestBed.get(UserService);
  //   expect(UserServiceMock.getAllUsers).toHaveBeenCalled();
  // });

});
