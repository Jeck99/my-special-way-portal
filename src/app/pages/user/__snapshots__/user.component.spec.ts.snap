// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`user component should render component as described in snapshot 1`] = `
<app-user
  dataSource={[Function MatTableDataSource]}
  dialog={[Function Object]}
  displayedColumns={[Function Array]}
  resultsLength="0"
  sort={[Function MatSort]}
  table={[Function ElementRef]}
  userService={[Function Object]}
>
  <div
    class="main-content"
  >
    <div
      class="container-fluid"
    >
      <div
        class="header-menu-container msw-table-tools-header"
      >
        <div
          class="example-header"
        >
          <mat-form-field
            class="search-user"
          >
            <input
              matinput=""
              placeholder="פילטר"
            />
          </mat-form-field>
        </div>
        <div
          class="example-header"
        >
          <button
            color="primary"
            mat-raised-button=""
          >
            הוספת משתמש חדש
          </button>
        </div>
      </div>
      <div
        class="row"
      >
        <div
          class="col-lg-12 col-md-12 col-sm-12"
        >
          <mat-table
            class="msw-table"
            matsort=""
            matsortactive="firstName"
            matsortdirection="asc"
            matsortdisableclear=""
          >
            
            
            
            
            
            
            
            
            
            
            
            
            
            
            
            
            
            
            
            --&gt; 
            
            
            
            
          </mat-table>
          <mat-paginator />
        </div>
      </div>
    </div>
  </div>
</app-user>
`;
